File: so_chi_tieu_app/lib/models/recurring_transaction.dart
// lib/models/recurring_transaction.dart

import 'category.dart';

enum RecurringFrequency { DAILY, WEEKLY, MONTHLY, YEARLY }

class RecurringTransaction {
  final int id;
  final double amount;
  final TransactionType type;
  final String? description;
  final int categoryId;
  final String? categoryName;
  final String? categoryIcon;
  final RecurringFrequency frequency;
  final DateTime startDate;
  final DateTime? endDate;
  final DateTime? nextExecutionDate;
  final bool isActive;
  final DateTime? createdAt;

  RecurringTransaction({
    required this.id,
    required this.amount,
    required this.type,
    this.description,
    required this.categoryId,
    this.categoryName,
    this.categoryIcon,
    required this.frequency,
    required this.startDate,
    this.endDate,
    this.nextExecutionDate,
    this.isActive = true,
    this.createdAt,
  });

  factory RecurringTransaction.fromJson(Map<String, dynamic> json) {
    return RecurringTransaction(
      id: json['id'],
      amount: (json['amount'] as num).toDouble(),
      type: json['type'] == 'INCOME' 
          ? TransactionType.INCOME 
          : TransactionType.EXPENSE,
      description: json['description'],
      categoryId: json['categoryId'],
      categoryName: json['categoryName'],
      categoryIcon: json['categoryIcon'],
      frequency: _parseFrequency(json['frequency']),
      startDate: DateTime.parse(json['startDate']),
      endDate: json['endDate'] != null ? DateTime.parse(json['endDate']) : null,
      nextExecutionDate: json['nextExecutionDate'] != null 
          ? DateTime.parse(json['nextExecutionDate']) 
          : null,
      isActive: json['isActive'] ?? true,
      createdAt: json['createdAt'] != null 
          ? DateTime.parse(json['createdAt']) 
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'amount': amount,
      'type': type == TransactionType.INCOME ? 'INCOME' : 'EXPENSE',
      'description': description,
      'categoryId': categoryId,
      'frequency': frequency.name,
      'startDate': startDate.toIso8601String().split('T')[0],
      'endDate': endDate?.toIso8601String().split('T')[0],
      'isActive': isActive,
    };
  }
  
  static RecurringFrequency _parseFrequency(String? value) {
    switch (value) {
      case 'DAILY':
        return RecurringFrequency.DAILY;
      case 'WEEKLY':
        return RecurringFrequency.WEEKLY;
      case 'MONTHLY':
        return RecurringFrequency.MONTHLY;
      case 'YEARLY':
        return RecurringFrequency.YEARLY;
      default:
        return RecurringFrequency.MONTHLY;
    }
  }
  
  String get frequencyDisplay {
    switch (frequency) {
      case RecurringFrequency.DAILY:
        return 'Hàng ngày';
      case RecurringFrequency.WEEKLY:
        return 'Hàng tuần';
      case RecurringFrequency.MONTHLY:
        return 'Hàng tháng';
      case RecurringFrequency.YEARLY:
        return 'Hàng năm';
    }
  }
}

File: src/main/java/com/thotran/sochitieu/entity/RecurringTransaction.java
package com.thotran.sochitieu.entity;

import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

/**
 * Entity đại diện cho giao dịch lặp lại.
 * VD: Lương hàng tháng, tiền nhà, học phí, Netflix subscription...
 */
@Entity
@Table(name = "recurring_transactions")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class RecurringTransaction {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    // Số tiền mỗi lần
    @Column(nullable = false, precision = 15, scale = 2)
    private BigDecimal amount;
    
    // Loại giao dịch: INCOME hoặc EXPENSE
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TransactionType type;
    
    // Mô tả (VD: "Lương tháng", "Tiền nhà")
    private String description;
    
    // Tần suất lặp lại
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private RecurringFrequency frequency;
    
    // Ngày bắt đầu
    @Column(name = "start_date", nullable = false)
    private LocalDate startDate;
    
    // Ngày kết thúc (nullable = vô thời hạn)
    @Column(name = "end_date")
    private LocalDate endDate;
    
    // Ngày đến hạn tiếp theo
    @Column(name = "next_due_date", nullable = false)
    private LocalDate nextDueDate;
    
    // Đang hoạt động?
    @Column(name = "is_active")
    @Builder.Default
    private Boolean isActive = true;
    
    // Thuộc danh mục nào
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    @ToString.Exclude
    private Category category;
    
    // Thuộc user nào
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    @ToString.Exclude
    private User user;
    
    @Column(name = "created_at")
    private LocalDateTime createdAt;
    
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
    
    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        if (nextDueDate == null) {
            nextDueDate = startDate;
        }
    }
    
    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
    
    /**
     * Tính ngày đến hạn tiếp theo dựa trên frequency
     */
    public LocalDate calculateNextDueDate() {
        return switch (frequency) {
            case DAILY -> nextDueDate.plusDays(1);
            case WEEKLY -> nextDueDate.plusWeeks(1);
            case MONTHLY -> nextDueDate.plusMonths(1);
            case YEARLY -> nextDueDate.plusYears(1);
        };
    }
}
